api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java

```diff
@@ -109,6 +109,10 @@ public Expression bind(StructType struct, boolean caseSensitive) {
       return bindUnaryOperation(bound);
     }
 
+    if (bound.type().isNestedType()) {
```

Comment:
Wondering if this is necessary.  Today will hit ValidationException when it tries to make a Literal?  

If necessary, we can use "Preconditions".

Reply:
If the predicate value is a complex type, for example
```
    Float[] value = {1.23f, 2.34f};
    UnboundPredicate<Float[]> lt = new UnboundPredicate<>(LT, ref("list"), value);
```
The Exception will be thrown at `Literals.from(value)` at this [line](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java#L40)

If the predicate value is a primitive, for example

```
    UnboundPredicate lt = new UnboundPredicate<>(LT, ref("list"), 1.23f);
```
It will throw Exception at this [line](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/expressions/UnboundPredicate.java#L116). So this `if (bound.type().isNestedType())` might be needed when a primitive value is accidentally used for Predicate on a nested type. 

I changed this to `Preconditions`.

Reply:
OK , makes sense to be more up front in what we support.